File: backend/src/agent_compiler/models/ir.py
"""Intermediate Representation (IR) schema for agent flows.

This module defines the Pydantic models for validating flow definitions,
including nodes, edges, and the overall flow structure.
"""

from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator, model_validator


class NodeType(str, Enum):
    """Supported node types in the flow."""

    LLM = "LLM"
    TOOL = "Tool"
    ROUTER = "Router"
    RETRIEVER = "Retriever"
    MEMORY = "Memory"
    OUTPUT = "Output"
    ERROR = "Error"  # Special node for error handling
    PARALLEL = "Parallel"
    JOIN = "Join"


class EngineType(str, Enum):
    """Supported execution engines."""

    LANGCHAIN = "langchain"
    LLAMAINDEX = "llamaindex"
    AUTO = "auto"


class LLMProvider(str, Enum):
    """Supported LLM providers."""

    OPENAI = "openai"
    GEMINI = "gemini"
    ANTHROPIC = "anthropic"
    AUTO = "auto"  # infer from model name (legacy fallback)


class SchemaKind(str, Enum):
    """Supported schema reference kinds."""

    JSON_SCHEMA = "json_schema"
    PYDANTIC = "pydantic"
    ZOD = "zod"


class SchemaRef(BaseModel):
    """Reference to an external schema artifact."""

    kind: SchemaKind = SchemaKind.JSON_SCHEMA
    ref: str = Field(..., min_length=1)


class RetrySpec(BaseModel):
    """Retry policy specification."""

    max_attempts: int = Field(default=2, ge=1, le=10)
    backoff_ms: int = Field(default=300, ge=0, le=60000)
    retry_on: list[str] = Field(default_factory=lambda: ["timeout", "rate_limit", "5xx"])
    jitter: bool = True


class FallbackSpec(BaseModel):
    """Fallback chain specification."""

    llm_chain: list[dict[str, Any]] = Field(default_factory=list)
    tool_fallbacks: dict[str, list[str]] = Field(default_factory=dict)


class LLMParams(BaseModel):
    """Parameters for LLM nodes."""

    provider: LLMProvider = LLMProvider.AUTO
    model: str = "gpt-3.5-turbo"
    temperature: float = Field(default=0.7, ge=0.0, le=2.0)
    system_prompt: str | None = None
    prompt_template: str = "{input}"
    max_tokens: int | None = None
    engine: EngineType | None = None
    is_start: bool = False
    # Retry & timeout
    retry_count: int = Field(default=0, ge=0, le=10, description="Number of retries on failure")
    retry_delay: float = Field(default=1.0, ge=0.0, description="Seconds between retries")
    timeout_seconds: float | None = Field(default=None, description="Max seconds per attempt")


class ToolParams(BaseModel):
    """Parameters for Tool nodes."""

    tool_name: str
    tool_config: dict[str, Any] = Field(default_factory=dict)
    engine: EngineType | None = None
    is_start: bool = False
    # Retry & timeout
    retry_count: int = Field(default=0, ge=0, le=10, description="Number of retries on failure")
    retry_delay: float = Field(default=1.0, ge=0.0, description="Seconds between retries")
    timeout_seconds: float | None = Field(default=None, description="Max seconds per attempt")
    retries: RetrySpec | None = None
    fallbacks: FallbackSpec | None = None
    input_schema: SchemaRef | None = None
    output_schema: SchemaRef | None = None


class RouterGuardMode(str, Enum):
    """Guard modes for Router nodes."""

    NONE = "none"  # Standard routing (default)
    RETRIEVAL = "retrieval"  # Guard based on retrieval evidence


class RouterGuardConfig(BaseModel):
    """Configuration for Router guard mode."""

    min_docs: int = Field(default=1, ge=0, description="Minimum documents required")
    min_top_score: float = Field(
        default=0.3, ge=0.0, le=1.0, description="Minimum score for top document"
    )
    else_branch: str = Field(
        default="abstain", description="Branch to take when guard fails"
    )


class RouterParams(BaseModel):
    """Parameters for Router nodes."""

    routes: dict[str, str] = Field(
        default_factory=dict,
        description="Mapping of condition -> target node ID",
    )
    default_route: str | None = None
    engine: EngineType | None = None
    is_start: bool = False

    # Guard mode for deterministic routing based on retrieval evidence
    guard_mode: RouterGuardMode = Field(
        default=RouterGuardMode.NONE,
        description="Guard mode: 'none' for standard routing, 'retrieval' for evidence-based",
    )
    guard_config: RouterGuardConfig = Field(
        default_factory=RouterGuardConfig,
        description="Configuration for guard mode evaluation",
    )


class RetrieverParams(BaseModel):
    """Parameters for Retriever nodes."""

    query_template: str = "{input}"
    top_k: int = Field(default=5, ge=1, le=100)
    index_name: str | None = None
    index_config: dict[str, Any] = Field(default_factory=dict)
    engine: EngineType | None = None
    is_start: bool = False
    retries: RetrySpec | None = None


class MemoryParams(BaseModel):
    """Parameters for Memory nodes."""

    memory_type: Literal["buffer", "summary", "vector"] = "buffer"
    max_tokens: int = 2000
    engine: EngineType | None = None
    is_start: bool = False


class OutputParams(BaseModel):
    """Parameters for Output nodes."""

    output_template: str = "{result}"
    format: Literal["text", "json", "markdown"] = "text"
    is_start: bool = False


class ErrorParams(BaseModel):
    """Parameters for Error nodes."""

    error_template: str = "An error occurred while processing this request."
    is_start: bool = False


class ParallelParams(BaseModel):
    """Parameters for Parallel fan-out nodes."""

    mode: Literal["broadcast"] = "broadcast"
    is_start: bool = False


class JoinParams(BaseModel):
    """Parameters for Join nodes."""

    strategy: Literal["array", "dict", "last_non_null"] = "array"
    is_start: bool = False


NodeParams = (
    LLMParams
    | ToolParams
    | RouterParams
    | RetrieverParams
    | MemoryParams
    | OutputParams
    | ErrorParams
    | ParallelParams
    | JoinParams
)


class Node(BaseModel):
    """A node in the flow graph."""

    id: str = Field(..., min_length=1)
    type: NodeType
    name: str = Field(..., min_length=1)
    params: dict[str, Any] = Field(default_factory=dict)

    @model_validator(mode="after")
    def validate_params(self) -> "Node":
        """Validate params based on node type."""
        param_classes: dict[NodeType, type[BaseModel]] = {
            NodeType.LLM: LLMParams,
            NodeType.TOOL: ToolParams,
            NodeType.ROUTER: RouterParams,
            NodeType.RETRIEVER: RetrieverParams,
            NodeType.MEMORY: MemoryParams,
            NodeType.OUTPUT: OutputParams,
            NodeType.ERROR: ErrorParams,
            NodeType.PARALLEL: ParallelParams,
            NodeType.JOIN: JoinParams,
        }
        param_class = param_classes.get(self.type)
        if param_class:
            param_class.model_validate(self.params)
        return self

    def get_typed_params(self) -> NodeParams:
        """Get params as typed object."""
        param_classes: dict[NodeType, type[BaseModel]] = {
            NodeType.LLM: LLMParams,
            NodeType.TOOL: ToolParams,
            NodeType.ROUTER: RouterParams,
            NodeType.RETRIEVER: RetrieverParams,
            NodeType.MEMORY: MemoryParams,
            NodeType.OUTPUT: OutputParams,
            NodeType.ERROR: ErrorParams,
            NodeType.PARALLEL: ParallelParams,
            NodeType.JOIN: JoinParams,
        }
        param_class = param_classes[self.type]
        return param_class.model_validate(self.params)


class Edge(BaseModel):
    """An edge connecting two nodes."""

    source: str = Field(..., min_length=1)
    target: str = Field(..., min_length=1)
    condition: str | None = None

    @field_validator("target")
    @classmethod
    def target_differs_from_source(cls, v: str, info) -> str:
        """Ensure target is different from source."""
        if info.data.get("source") == v:
            raise ValueError("Self-loops are not allowed")
        return v


class Flow(BaseModel):
    """Flow metadata."""

    id: str = Field(..., min_length=1)
    name: str = Field(..., min_length=1)
    version: str = "1.0.0"
    engine_preference: EngineType = EngineType.LANGCHAIN
    description: str = ""


class FlowIR(BaseModel):
    """Complete flow Intermediate Representation."""

    ir_version: Literal["2"] = "2"
    flow: Flow
    nodes: list[Node] = Field(..., min_length=1)
    edges: list[Edge] = Field(default_factory=list)

    @model_validator(mode="after")
    def validate_flow_structure(self) -> "FlowIR":
        """Validate the flow structure: DAG, start node, node references."""
        node_ids = {node.id for node in self.nodes}

        # Validate edge references
        for edge in self.edges:
            if edge.source not in node_ids:
                raise ValueError(f"Edge source '{edge.source}' references unknown node")
            if edge.target not in node_ids:
                raise ValueError(f"Edge target '{edge.target}' references unknown node")

        # Find start nodes
        nodes_with_incoming = {edge.target for edge in self.edges}
        marked_start_nodes = [
            node for node in self.nodes if node.params.get("is_start", False)
        ]
        implicit_start_nodes = [
            node for node in self.nodes if node.id not in nodes_with_incoming
        ]

        if marked_start_nodes:
            if len(marked_start_nodes) > 1:
                raise ValueError(
                    f"Multiple nodes marked as start: {[n.id for n in marked_start_nodes]}"
                )
            self._start_node_id = marked_start_nodes[0].id
        elif len(implicit_start_nodes) == 1:
            self._start_node_id = implicit_start_nodes[0].id
        elif len(implicit_start_nodes) > 1:
            raise ValueError(
                f"Ambiguous start nodes (no incoming edges): {[n.id for n in implicit_start_nodes]}. "
                "Mark one with is_start=true"
            )
        else:
            raise ValueError("No start node found")

        # Validate DAG (no cycles)
        self._validate_dag(node_ids)

        return self

    def _validate_dag(self, node_ids: set[str]) -> None:
        """Validate that the graph is a DAG (no cycles)."""
        adjacency: dict[str, list[str]] = {nid: [] for nid in node_ids}
        for edge in self.edges:
            adjacency[edge.source].append(edge.target)

        # Kahn's algorithm for topological sort / cycle detection
        in_degree: dict[str, int] = {nid: 0 for nid in node_ids}
        for edge in self.edges:
            in_degree[edge.target] += 1

        queue = [nid for nid, deg in in_degree.items() if deg == 0]
        visited = 0

        while queue:
            node = queue.pop(0)
            visited += 1
            for neighbor in adjacency[node]:
                in_degree[neighbor] -= 1
                if in_degree[neighbor] == 0:
                    queue.append(neighbor)

        if visited != len(node_ids):
            raise ValueError("Flow contains a cycle - only DAGs are supported in MVP")

    @property
    def start_node_id(self) -> str:
        """Get the ID of the start node."""
        return getattr(self, "_start_node_id", self.nodes[0].id)

    def get_node(self, node_id: str) -> Node | None:
        """Get a node by ID."""
        for node in self.nodes:
            if node.id == node_id:
                return node
        return None

    def get_topological_order(self) -> list[str]:
        """Get nodes in topological order."""
        node_ids = {node.id for node in self.nodes}
        adjacency: dict[str, list[str]] = {nid: [] for nid in node_ids}
        for edge in self.edges:
            adjacency[edge.source].append(edge.target)

        in_degree: dict[str, int] = {nid: 0 for nid in node_ids}
        for edge in self.edges:
            in_degree[edge.target] += 1

        queue = [nid for nid, deg in in_degree.items() if deg == 0]
        result: list[str] = []

        while queue:
            # Sort for deterministic ordering
            queue.sort()
            node = queue.pop(0)
            result.append(node)
            for neighbor in adjacency[node]:
                in_degree[neighbor] -= 1
                if in_degree[neighbor] == 0:
                    queue.append(neighbor)

        return result

    def get_successors(self, node_id: str) -> list[str]:
        """Get successor node IDs for a given node."""
        return [edge.target for edge in self.edges if edge.source == node_id]

    def get_predecessors(self, node_id: str) -> list[str]:
        """Get predecessor node IDs for a given node."""
        return [edge.source for edge in self.edges if edge.target == node_id]


# =============================================================================
# IR Parsing
# =============================================================================


def parse_ir(data: dict) -> "FlowIRv2":
    """Parse raw IR data as v2 only.

    Raises:
        ValueError: If the payload is not IR v2.
    """
    from agent_compiler.models.ir_v2 import FlowIRv2
    from agent_compiler.models.ir_v2_to_v2_1 import migrate_v2_to_v2_1_defaults

    version = data.get("ir_version")
    if version != "2":
        raise ValueError("Only ir_version='2' is supported.")
    return migrate_v2_to_v2_1_defaults(FlowIRv2.model_validate(data))

File: frontend/src/types/ir.ts
// IR (Intermediate Representation) Types for Agent Compiler
// Canonical FE flow types aligned to backend IR v2.

export const IR_VERSION = "2" as const;

export type EnginePreference = "langchain" | "llamaindex" | "auto";

// Node types must be UPPERCASE to match backend
export type NodeType = "LLM" | "Tool" | "Router" | "Retriever" | "Memory" | "Output" | "Error" | "Parallel" | "Join";

export interface NodePosition {
  x: number;
  y: number;
}

// Base params that all nodes share
interface BaseNodeParams {
  engine?: EnginePreference | null;
  is_start?: boolean;
}

// LLM providers
export type LLMProvider = "auto" | "openai" | "gemini" | "anthropic";

// Node-specific parameter types - matching backend exactly
export interface LLMNodeParams extends BaseNodeParams {
  provider: LLMProvider;
  model: string;
  temperature: number;
  system_prompt?: string | null;
  prompt_template: string;
  max_tokens?: number | null;
  retry_count?: number;
  retry_delay?: number;
  timeout_seconds?: number | null;
}

export interface ToolNodeParams extends BaseNodeParams {
  tool_name: string;
  tool_config: Record<string, unknown>;
  retry_count?: number;
  retry_delay?: number;
  timeout_seconds?: number | null;
}

export type RouterMode = "llm" | "guard";

export interface RouterNodeParams extends BaseNodeParams {
  routes: Record<string, string>; // condition -> target node ID
  default_route?: string | null;
  // Guard mode fields
  mode?: RouterMode;
  min_docs?: number;
  min_top_score?: number;
  grounded_branch?: string | null;
  abstain_branch?: string | null;
}

export interface RetrieverNodeParams extends BaseNodeParams {
  query_template: string;
  top_k: number;
  index_name?: string | null;
  index_config: Record<string, unknown>;
}

export interface MemoryNodeParams extends BaseNodeParams {
  memory_type: "buffer" | "summary" | "vector";
  max_tokens: number;
}

export interface OutputNodeParams {
  output_template: string;
  format: "text" | "json" | "markdown";
  is_start?: boolean;
}

export interface ErrorNodeParams extends BaseNodeParams {
  error_template?: string;
}

export interface ParallelNodeParams extends BaseNodeParams {
  mode?: "broadcast";
}

export interface JoinNodeParams extends BaseNodeParams {
  strategy?: "array" | "dict" | "last_non_null";
}

export type NodeParams =
  | LLMNodeParams
  | ToolNodeParams
  | RouterNodeParams
  | RetrieverNodeParams
  | MemoryNodeParams
  | OutputNodeParams
  | ErrorNodeParams
  | ParallelNodeParams
  | JoinNodeParams;

// Node as sent to/from backend (no position - that's frontend only)
export interface FlowNodeBackend {
  id: string;
  type: NodeType;
  name: string;
  params: NodeParams;
}

// Node with frontend-specific position data
export interface FlowNode extends FlowNodeBackend {
  position: NodePosition;
}

// Edge structure - uses "condition" not "label"
export interface FlowEdge {
  source: string;
  target: string;
  condition?: string | null;
}

// Flow metadata
export interface FlowMeta {
  id: string;
  name: string;
  version: string;
  engine_preference: EnginePreference;
  description: string;
}

// Frontend flow with positions
export interface Flow extends FlowMeta {
  ir_version: string;
  // In v2, nodes/edges are a derived canvas view from agents.graph.
  nodes: FlowNode[];
  edges: FlowEdge[];
  created_at?: string;
  updated_at?: string;
  // Multi-agent fields stored in the canonical backend IR.
  agents?: import("./agents").AgentSpec[];
  handoffs?: import("./agents").HandoffRule[];
  entrypoints?: import("./agents").EntrypointSpec[];
  resources?: import("./agents").ResourceRegistry;
  policies?: import("./agents").PolicySpec;
}

// API Response types
export interface FlowListItem {
  id: string;
  name: string;
  version: string;
  description: string;
  engine_preference: string;
  created_at: string;
  updated_at: string;
}

export interface FlowDetailResponse extends FlowListItem {
  ir: import("./agents").FlowIRv2;
}

// Run types
export type RunStatus = "pending" | "running" | "completed" | "failed" | "cancelled";
export type StepStatus = "pending" | "running" | "completed" | "failed" | "skipped";

// Retrieved document from Retriever step
export interface RetrievedDoc {
  doc_id: string;
  source: string;
  snippet: string;
  score: number;
}

// Citation in output
export interface Citation {
  doc_id: string;
  source: string;
  text: string;
}

// Guard decision info
export interface GuardDecision {
  decision: "grounded" | "abstain";
  threshold: number;
  top_score: number;
  reason?: string;
}

export interface TimelineStep {
  step_id: string;
  node_id: string;
  node_type: string;
  status: StepStatus;
  input: Record<string, unknown>;
  output: Record<string, unknown> | null;
  meta: Record<string, unknown>;
  error_message: string | null;
  started_at: string | null;
  finished_at: string | null;
  // Enhanced fields for Guard/Citations
  retrieved_docs?: RetrievedDoc[];
  guard_decision?: GuardDecision;
  citations?: Citation[];
  // Token tracking
  tokens?: { input?: number | null; output?: number | null; total?: number | null } | null;
  model_name?: string | null;
  // Multi-agent fields
  agent_id?: string | null;
  depth?: number;
  parent_step_id?: string | null;
}

export interface Run {
  id: string;
  flow_id: string;
  status: RunStatus;
  input: Record<string, unknown>;
  output: Record<string, unknown> | null;
  error_message: string | null;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
  timeline: TimelineStep[];
  // Replay related fields
  original_run_id?: string | null;
  replay_mode?: "exact" | "mocked" | null;
  citations?: Citation[];
  // Multi-agent fields
  entrypoint?: string;
  agent_events?: import("./agents").AgentEvent[];
}

export interface RunListItem {
  id: string;
  flow_id: string;
  status: RunStatus;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
}

// Default params for each node type - matching backend defaults
export const DEFAULT_NODE_PARAMS: Record<NodeType, NodeParams> = {
  LLM: {
    provider: "auto",
    model: "gpt-3.5-turbo",
    temperature: 0.7,
    system_prompt: null,
    prompt_template: "{input}",
    max_tokens: null,
    engine: null,
    is_start: false,
    retry_count: 0,
    retry_delay: 1.0,
    timeout_seconds: null,
  },
  Tool: {
    tool_name: "",
    tool_config: {},
    engine: null,
    is_start: false,
    retry_count: 0,
    retry_delay: 1.0,
    timeout_seconds: null,
  },
  Retriever: {
    query_template: "{input}",
    top_k: 5,
    index_name: null,
    index_config: {},
    engine: null,
    is_start: false,
  },
  Memory: {
    memory_type: "buffer",
    max_tokens: 2000,
    engine: null,
    is_start: false,
  },
  Router: {
    routes: {},
    default_route: null,
    engine: null,
    is_start: false,
    mode: "llm",
    min_docs: 1,
    min_top_score: 0.65,
    grounded_branch: null,
    abstain_branch: null,
  },
  Output: {
    output_template: "{result}",
    format: "text",
    is_start: false,
  },
  Error: {
    error_template: "An error occurred while processing this request.",
    is_start: false,
  },
  Parallel: {
    mode: "broadcast",
    is_start: false,
  },
  Join: {
    strategy: "array",
    is_start: false,
  },
};

export const NODE_TYPE_LABELS: Record<NodeType, string> = {
  LLM: "LLM",
  Tool: "Tool",
  Router: "Router",
  Retriever: "Retriever",
  Memory: "Memory",
  Output: "Output",
  Error: "Error",
  Parallel: "Parallel",
  Join: "Join",
};

export const NODE_TYPE_COLORS: Record<NodeType, string> = {
  LLM:       "#8b5cf6",   // --node-llm
  Tool:      "#f97316",   // --node-tool
  Router:    "#22c55e",   // --node-router
  Retriever: "#3b82f6",   // --node-retriever
  Memory:    "#ec4899",   // --node-memory
  Output:    "#06b6d4",   // --node-output
  Error:     "#ef4444",   // --accent-error
  Parallel:  "#eab308",   // --node-trigger
  Join:      "#22d3ee",   // --accent-primary
};

// Helper to convert lowercase type to uppercase
export function normalizeNodeType(type: string): NodeType {
  // Handle case variations
  const mapping: Record<string, NodeType> = {
    "llm": "LLM",
    "tool": "Tool",
    "router": "Router",
    "retriever": "Retriever",
    "memory": "Memory",
    "output": "Output",
    "error": "Error",
    "parallel": "Parallel",
    "join": "Join",
  };
  return mapping[type.toLowerCase()] || "LLM";
}
